[TUTORIAL] Chatbot-tutorial (#122)
huggingface/hf-endpoints-documentation: docs/source/tutorials/chat_bot.mdx

# Build and deploy your own chat application

> This tutorial will guide you from end to end on how to deploy your own chat application using Hugging Face Inference Endpoints. We will use Gradio to create a chat interface and OpenAI client to connect to the Inference Endpoint.

Reviewer: ```suggestion
This tutorial will guide you from end to end on how to deploy your own chat application using Hugging Face Inference Endpoints. We will use Gradio to create a chat interface and an OpenAI client to connect to the Inference Endpoint.
```

<Tip>

> This Tutorial uses Python, but your client can be any language that can make HTTP requests. Inference Endpoints uses the **OpenAI API format**, so you can use any [OpenAI client](https://platform.openai.com/docs/libraries) to connect to them, in languages like JavaScript, Java, and Go.

Reviewer 2: ```suggestion
This Tutorial uses Python, but your client can be any language that can make HTTP requests. Inference Endpoints uses the **OpenAI Chat Completions format**, so you can use any [OpenAI client](https://platform.openai.com/docs/libraries) to connect to them, in languages like JavaScript, Java, and Go.
```

</Tip>

## Create your endpoint

First, we need to create an Inference Endpoint for a model that can chat. 

Start by navigating to the Inference Endpoints UI, and once you have logged in you should see a button for creating a new Inference
Endpoint, and a small greeting prompting you to create your first endpoint. Click the "New" button.

![new-button](https://raw.githubusercontent.com/huggingface/hf-endpoints-documentation/update/large-rewrite/assets/quick_start/1-new-button.png)

From there you'll be directed to the catalog. The Model Catalog consists of popular models which have tuned configurations to work just as one-click
deploys. You can filter by name, task, price of the hardware and much more.

![catalog](https://raw.githubusercontent.com/huggingface/hf-endpoints-documentation/update/large-rewrite/assets/quick_start/2-catalog.png)

> In this example let's deploy the [meta-llama/Llama-3.2-3B-Instruct](https://huggingface.co/meta-llama/Llama-3.2-3B-Instruct) model. You can find

Reviewer 2: I'd recommend using Qwen3 / SmolLM3 - a lot of people face gating issues w/ Meta/ Google so better to choose a model that is ungated for the guide

Author: Nice. I'll do that!

Now our Inference Endpoint is initializing, which usually takes about 3-5 minutes. If you want to can alow browser notifications which will give you a
ping once the endpoint reaches a running state.

![init](https://raw.githubusercontent.com/huggingface/hf-endpoints-documentation/update/large-rewrite/assets/quick_start/5-init.png)

## Get your endpoint details

Now that we've created our endpoint, we need to get the endpoint details which we can find on the [endpoints unique page](https://endpoints.huggingface.co/). We will need the following details:

- The base URL of the endpoint plus the version of the OpenAI API (e.g. `https://<endpoint-name>.endpoints.huggingface.cloud/v1/`)
- The name of the endpoint to use (e.g. `meta-llama/Llama-3.2-3B-Instruct`)
- The token to use for authentication (e.g. `hf_<token>`)

![endpoint-details](https://raw.githubusercontent.com/huggingface/hf-endpoints-documentation/update/chatbot-tutorial/assets/tutorials/chatbot/endpoint-page.png)

We can find the token on the [settings page](https://huggingface.co/settings/tokens) which is accessible from the top right corner of the endpoint page.

> ## Deploy in a few lines of code

Reviewer 2: maybe just above this it's worth showing that one can test the endpoint in the playground on the same page itself.

The easiest way to deploy a chat application with [Gradio](https://gradio.app/) is to use the convenient `load_chat` method. This abstracts everything away and you can have a working chat application quickly.

```python
import os

import gradio as gr

gr.load_chat(
    base_url="<endpoint-url>/v1/", # Replace with your endpoint URL + version
    model="endpoint-name", # Replace with your endpoint name
    token=os.getenv("HF_TOKEN"), # Replace with your token
).launch()
```

The `load_chat` method won't cater for your production needs, but it's a great way to get started and test your application. 

> ## Build your own custom chat application

Reviewer 2: slightly controversial but maybe worth showcasing *just* the OpenAI client here (it's way more used compared to HFH IC anyway)

Author: Thought about it and I'm not going to remove the others. Mainly just because they're in tabs and OpenAI is first, so I don't feel like there's a net benefit to dropping them.

Choose your preferred method for connecting to Inference Endpoints:

<hfoptions id="chat-implementation">
<hfoption id="hf-client">

**Using Hugging Face InferenceClient**

First, install the required dependencies:

```bash
pip install gradio huggingface-hub
```

The Hugging Face InferenceClient provides a clean interface that's compatible with the OpenAI API format:

```python
import os
import gradio as gr
from huggingface_hub import InferenceClient

# Initialize the Hugging Face InferenceClient
client = InferenceClient(
    base_url="<endpoint-url>/v1/",  # Replace with your endpoint URL
    token=os.getenv("HF_TOKEN")  # Use environment variable for security
)

def chat_with_hf_client(message, history):
    # Convert Gradio history to messages format
    messages = [{"role": msg["role"], "content": msg["content"]} for msg in history]
    
    # Add the current message
    messages.append({"role": "user", "content": message})
    
    # Create chat completion
    chat_completion = client.chat.completions.create(

>         model="endpoint-name",  # Use your model name

Author: In my test app I use the endpoint name like `smollm2-1-7b-instruct-ljn` which I'm calling 'endpoint-name' in the tutorial because model name could be `SmolLM2-1-7b`

    return chat_completion.choices[0].message.content

# Create the Gradio interface
demo = gr.ChatInterface(
    fn=chat_with_hf_client,
    type="messages",
    title="Custom Chat with Inference Endpoints",
    examples=["What is deep learning?", "Explain neural networks", "How does AI work?"]
)

if __name__ == "__main__":
    demo.launch()
```

</hfoption>
<hfoption id="openai-client">

**Using OpenAI Client**

First, install the required dependencies:
```bash
pip install gradio openai
```

Here's a basic chat function using the OpenAI client:

```python
import os
import gradio as gr
from openai import OpenAI

# Initialize the OpenAI client with your Inference Endpoint
client = OpenAI(
    base_url="<endpoint-url>/v1/",  # Replace with your endpoint URL
    api_key=os.getenv("HF_TOKEN")  # Use environment variable for security
)

def chat_with_openai(message, history):

    # Convert Gradio history to OpenAI format
    messages = [{"role": msg["role"], "content": msg["content"]} for msg in history]
    
    # Add the current message
    messages.append({"role": "user", "content": message})
    
    # Create chat completion
    chat_completion = client.chat.completions.create(

>         model="endpoint-name",  # Use your model name

Reviewer: same here

}

def chat_with_requests(message, history):
    # Convert Gradio history to messages format
    messages = [{"role": msg["role"], "content": msg["content"]} for msg in history]
    
    # Add the current message
    messages.append({"role": "user", "content": message})
    
    # Prepare the payload using the Messages API format
    payload = {
        "model": "endpoint-name",  # Use your model name
        "messages": messages,
        "max_tokens": 150,
        "temperature": 0.7,
        "stream": False
    }
    
    # Make the request
    response = requests.post(API_URL, headers=headers, json=payload)
    response.raise_for_status()
    
    # Parse the response
    result = response.json()
    return result["choices"][0]["message"]["content"]

# Create the Gradio interface
demo = gr.ChatInterface(
    fn=chat_with_requests,
    type="messages",
    title="Custom Chat with Inference Endpoints",
    examples=["What is deep learning?", "Explain neural networks", "How does AI work?"]
)

if __name__ == "__main__":
    demo.launch()
```

</hfoption>
</hfoptions>

## Adding Streaming Support

> For a better user experience, you can implement streaming responses. This will require us to hand the messages and `yield` them to the client.

Reviewer: maybe this is a typo?
```suggestion
For a better user experience, you can implement streaming responses. This will require us to handle the messages and `yield` them to the client.
```
